### PR TITLE
Make WIP prefixes case insensitive, e.g. allow `Draft` as a WIP prefix

### LINKED ISSUE #19779: WORK_IN_PROGRESS_PREFIXES=Draft: still allows to merge "Draft" merge requests
### Description

Steps to reproduce: Put this into `custom/conf/app.ini`:
```toml
[repository.pull-request]
WORK_IN_PROGRESS_PREFIXES=Draft:,[Draft],WIP:,[WIP]
```
Then create a new pull request, add `Draft: ` at the beginning (for example by clicking the suggested link at the top that correctly shows use `Draft:` to prevent accidentally merging). It will create a pull request, that however can be merged as usual.

If you put `WIP: ` at the beginning, it will not allow you to merge it.

### Gitea Version

a9af93cb216e9b5f4d1841d18593589143fbdc9c

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

2.32.0

### Operating System

macOS Monterey

### How are you running Gitea?

I built it from source from git.

### Database

SQLite

### PR DESCRIPTION
**Pull request by certik**
There is a bug in WIP prefix matching which results in only all uppercase prefixes being matched.

The issue was that only the actual title was converted to uppercase, but
not the prefix as specified in `WORK_IN_PROGRESS_PREFIXES`. As a result,
the following did not work:

    WORK_IN_PROGRESS_PREFIXES=Draft:,[Draft],WIP:,[WIP]

One possible workaround was:

    WORK_IN_PROGRESS_PREFIXES=DRAFT:,[DRAFT],WIP:,[WIP]

Then indeed one could use `Draft` (as well as `DRAFT`) in the title.
However, the link `Start the title with DRAFT: to prevent the pull request
from being merged accidentally.` showed the suggestion in uppercase; so
it is not possible to show it as `Draft`. This PR fixes it, and allows
to use `Draft` in `WORK_IN_PROGRESS_PREFIXES`.

Fixes #19779.

### CODE REVIEW

**Comment by zeripath** [comment]
You also need to fix GetWorkInProgressPrefix() a few lines below that.

**Comment by zeripath** [comment]
In actuality I think the better solution would be to go to modules/setting/repository.go line 297 and uppercase the WorkInProgressPrefixes just after reading them from the configuration file.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
It's better to update the comment or document that is not case sensitive.

**Comment by certik** [comment]
> In actuality I think the better solution would be to go to modules/setting/repository.go line 297 and uppercase the WorkInProgressPrefixes just after reading them from the configuration file.

I think that would be equivalent to the workaround I mentioned: `WORK_IN_PROGRESS_PREFIXES=DRAFT:,[DRAFT],WIP:,[WIP]`, which works, but then Gitea puts `DRAFT` instead of `Draft` into the suggested prefix for Work in Progress. So I think a better solution is the one provided in this PR.


> You also need to fix GetWorkInProgressPrefix() a few lines below that.

Excellent point, fixed in 11754a123b4a4da732114efc4cba60a739fdd6a6. This is needed so that the "Remove Draft Prefix" button works. I tested it and it seems to work.

**Comment by zeripath** [comment]
Cool. Thanks...

It would be helpful to update:

https://github.com/go-gitea/gitea/blob/4266bd924bb8fff20271c17f8989b39b04ef6673/custom/conf/app.example.ini#L941

and

https://github.com/go-gitea/gitea/blob/4266bd924bb8fff20271c17f8989b39b04ef6673/docs/content/doc/advanced/config-cheat-sheet.en-us.md?plain=1#L90

To mention that these are matched in a case insensitive manner too. Then I will approve and we can get this merged.

**Comment by certik** [comment]
@zeripath I made the changes, good point. Let me know if there is anything else to be done.

File: custom/conf/app.example.ini [hunk 1]
```diff
@@ -938,8 +938,8 @@ PATH =
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;
-;; List of prefixes used in Pull Request title to mark them as Work In Progress
-;WORK_IN_PROGRESS_PREFIXES = WIP:,[WIP]
+;; List of prefixes used in Pull Request title to mark them as Work In Progress (matched in a case-insensitive manner)
+;WORK_IN_PROGRESS_PREFIXES = Draft:,[Draft],WIP:,[WIP]
```
Comment (zeripath) [inline custom/conf/app.example.ini]: You haven't changed the defaults in modules/setting/repository.go, therefore do not change the documented defaults. These have to match the real situation.

In reality I think changing the defaults would be a breaking change therefore we should not do that. So I'll write a suggestion and change them back.

```suggestion
;WORK_IN_PROGRESS_PREFIXES = WIP:,[WIP]
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 2]
```diff
@@ -86,8 +86,8 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 
 ### Repository - Pull Request (`repository.pull-request`)
 
-- `WORK_IN_PROGRESS_PREFIXES`: **WIP:,\[WIP\]**: List of prefixes used in Pull Request
- title to mark them as Work In Progress
+- `WORK_IN_PROGRESS_PREFIXES`: **Draft:,\[Draft\],WIP:,\[WIP\]**: List of prefixes used in Pull Request
```
Comment (zeripath) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]:
```suggestion
- `WORK_IN_PROGRESS_PREFIXES`: **WIP:,\[WIP\]**: List of prefixes used in Pull Request
```

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
make lgtm work

**Comment by zeripath** [comment]
I think it would be appropriate to backport this to 1.16.

If you would like to do this:

In your git repo

```sh
## Update your repo
git pull --all
 
## Checkout a backport branch
git checkout origin/release/v1.16 -b backport-19780-v1.16

## Cherry pick the above commit
git cherry-pick 8720f87

## and fix any conflicts...

## Adjust the commit message to make the first line after the title (and its blank line):
## Backport #19780 
git commit --amend

## Push it up
git push certik backport-19780-v1.16

## Then open a PR
xdg-open https://github.com/go-gitea/gitea/compare/release/"$VERSION"...certik:backport-19780-v1.16

## Put it on the v1.16.9 milestone with the label: kind/bug
## And then one of us will add the backport/done label to this.
```

File: custom/conf/app.example.ini (hunk 1 above)
Comment (certik) [reply custom/conf/app.example.ini]: Yes, that's fine. I agree we should not change the defaults in a PR like this.

**Comment by certik** [comment]
@zeripath thanks for fixing it up and merging. I backported it here: https://github.com/go-gitea/gitea/pull/19811, it seems I am not allowed to change milestones and labels, so you'll have to do that.

**Pull request merged**